### PR TITLE
Proper support for GIFs when using a RetainingDataSource

### PR DESCRIPTION
**Pull request by Crysis21**
## Motivation (required)
During the development of the Giphy app we choose to use fresco as our main image library. The application loads different types of image renditions, starting from low-quality images to high-quality ones. We wanted to be able to change the quality of a GIF displayed in a `DraweeView`.

When making a simple request to change the resource of a `DraweeView`, fresco will clear the current content, display the placeholder, then load the new resource. Even with the resources preloaded, there is a noticeable flickering.

Available solutions:
1. Use a low-res/hi-res schema for loading images. Unfortunately, in this schema, the low-res image is not animated.
2. Use a `RetainingDataSource`. While it's purpose was exactly what we were looking for, it turns out that this data source, does it's magic by lying to the `AbstractDataSource` code and keeping it in a forever `PROGRESS` state. That means, the datasource will never get to notify the listening controllers. This is the reason GIFs are not playing and controller callbacks are not fired using this `DataSource`.

## Test Plan (required)
Given a list of GIFs, by using a `RetainingDataSource` we should be able to replace the current playing view with a freshly loaded one, without the user notice any flickering or blank spaces between the 2 resouces.

I've created a set of test GIFS in the `ImageUriProvider` and I updated the `RetainingDataSourceSupplierFragment` to use those resouces, in order to demonstrate the new achieved behaviour.

**Demo video:**
https://youtu.be/A67X6Vn9VjY

### CODE REVIEW

**Comment by bogdantmm92** [comment]
Really good change! Thanks @Crysis21

**Comment by amarkovits** [comment]
Great solution.

File: fbcore/src/main/java/com/facebook/datasource/DataSource.java [hunk 1]
```diff
@@ -49,6 +49,13 @@
    */
   boolean hasResult();
 
+  /**
+   *
+   * @return true if any resources poured in the datasource should be sent to controllers. Usefull
+   * with a RetainingDataSource to loaded resources are displayed correctly.
+   */
+  boolean isReadyToPlay();
```
[the remark this answers is not in the document]
Comment (oprisnik) [reply fbcore/src/main/java/com/facebook/datasource/DataSource.java]: I suggest renaming this to something like `boolean hasMultipleResults()` and using that one instead.

[the remark this answers is not in the document]
Comment (Crysis21) [reply fbcore/src/main/java/com/facebook/datasource/DataSource.java]: 👍

File: drawee/src/main/java/com/facebook/drawee/controller/AbstractDraweeController.java [hunk 2]
```diff
@@ -532,7 +537,10 @@ private void onNewResultInternal(
           mDataSource = null;
           mSettableDraweeHierarchy.setImage(drawable, 1f, wasImmediate);
           getControllerListener().onFinalImageSet(id, getImageInfo(image), getAnimatable());
-          // IMPORTANT: do not execute any instance-specific code after this point
```
[the remark this answers is not in the document]
Comment (Crysis21) [reply drawee/src/main/java/com/facebook/drawee/controller/AbstractDraweeController.java]: ok, I'll add it

File: drawee/src/main/java/com/facebook/drawee/controller/AbstractDraweeController.java [hunk 3]
```diff
@@ -27,10 +26,14 @@
 import com.facebook.drawee.interfaces.SettableDraweeHierarchy;
 import com.facebook.imagepipeline.systrace.FrescoSystrace;
 import com.facebook.infer.annotation.ReturnsOwnership;
+
 import java.util.concurrent.Executor;
+
 import javax.annotation.Nullable;
 import javax.annotation.concurrent.NotThreadSafe;
 
+import static com.facebook.drawee.components.DraweeEventTracker.Event;
```
[the remark this answers is not in the document]
Comment (Crysis21) [reply drawee/src/main/java/com/facebook/drawee/controller/AbstractDraweeController.java]: AndroidStudio did that, imports optimisation was active. Will check for changes and revert them.

File: drawee/src/main/java/com/facebook/drawee/controller/AbstractDraweeController.java [hunk 4]
```diff
@@ -499,7 +503,8 @@ private void onNewResultInternal(
       @Nullable T image,
       float progress,
       boolean isFinished,
-      boolean wasImmediate) {
+      boolean wasImmediate,
+      boolean isReadyToPlay) {
```
[the remark this answers is not in the document]
Comment (Crysis21) [reply drawee/src/main/java/com/facebook/drawee/controller/AbstractDraweeController.java]: I renamed it to `deliverTempResult`. It's more accurate to what the parameter does.

**Comment by Crysis21** [comment]
I've applied the requested changes.

File: drawee/src/main/java/com/facebook/drawee/controller/AbstractDraweeController.java [hunk 5]
```diff
@@ -532,6 +534,10 @@ private void onNewResultInternal(
           mDataSource = null;
           mSettableDraweeHierarchy.setImage(drawable, 1f, wasImmediate);
           getControllerListener().onFinalImageSet(id, getImageInfo(image), getAnimatable());
+        } else if (deliverTempResult) {
+          logMessageAndImage("set_temporary_result @ onNewResult", image);
+          mSettableDraweeHierarchy.setImage(drawable, 1f, wasImmediate);
+          getControllerListener().onFinalImageSet(id, getImageInfo(image), getAnimatable());
           // IMPORTANT: do not execute any instance-specific code after this point
```
Comment (oprisnik) [inline drawee/src/main/java/com/facebook/drawee/controller/AbstractDraweeController.java]: This message should also be here for the original block. I'll just add it when I'm importing the change.

**Pull request closed without merging**